### PR TITLE
Replace manual fetch/XHR headers with first-class option

### PR DESCRIPTION
**Pull request by apasel422**
#756

***
<a href="https://pr-preview.s3.amazonaws.com/apasel422/attribution-reporting-api/pull/757.html" title="Last updated on Apr 19, 2023, 4:48 PM UTC (dfaf258)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/757/a82aaf0...apasel422:dfaf258.html" title="Last updated on Apr 19, 2023, 4:48 PM UTC (dfaf258)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (csharrison) [inline index.bs:143]: The existence of `source-and-trigger` makes me feel like an enum is not quite the right data structure to use here. WDYT about a dictionary (bikeshedding welcome):
```
fetch(<url>, {
  attributionReporting: {
    sourceEligibility: 'event',
    triggerEligibility: true,
  }
}
```
This sort of follows the advice here:
https://www.w3.org/TR/design-principles/#string-constants

"If you need to express a state which is a combination of properties, which might be expressed as a bitmask in another language, use a dictionary object instead. This object can be passed around as easily as a single bitmask value."

An added benefit is that an ARA-specific dictionary allows easier extensibility in the future.

Comment (apasel422) [reply index.bs:143]: I'm fine with that, though I'd prefer to use a boolean for `sourceEligibility` and call it `eventSourceEligible`. Otherwise, we have to deal with values other than `event`.

Comment (csharrison) [reply index.bs]: Perfect

**Review by csharrison — APPROVED** [review]

File: index.bs
Comment (apasel422) [inline index.bs:149]: Apparently dictionaries cannot be used as attributes, so this won't work.

https://webidl.spec.whatwg.org/#idl-attributes

File: index.bs
Comment (apasel422) [inline index.bs:145]:
```
Nullable dictionary type is forbidden as a dictionary member type.
```

https://webidl.spec.whatwg.org/#idl-dictionaries

**Comment by yoavweiss** [comment]
Is XHR a strong use case? Can't folks that care about this move to `fetch()`? (I'd have assumed most already use `fetch()`, but I could be wrong)

**Comment by apasel422** [comment]
> Is XHR a strong use case? Can't folks that care about this move to `fetch()`? (I'd have assumed most already use `fetch()`, but I could be wrong)

I don't feel strongly about supporting XHR (though the Chromium implementation does today via manual setting of the relevant header), but omitting it doesn't solve the optional dictionary problem for `fetch`.

Some other ideas:

1. Remove fine-grained configuration, and just have the following, where `null` means that the `Attribution-Reporting-Eligible` header is omitted but triggers are supported, `false` means that the header is set to the empty string and neither sources nor triggers are supported, and `true` means the header is set to `Attribution-Reporting-Eligible: event-source, trigger` and both are supported:
```webidl
partial dictionary RequestInit {
  boolean? attributionReportingEligible;
}
partial interface XMLHttpRequest {
  attribute boolean? attributionReportingEligible;
}
```
This makes `fetch` and `XHR` able to behave equivalently to ``, with the additional ability to disable registration 
entirely, which `` does not support, but obviously doesn't allow opting in to only triggers or only sources with the header set.

2. Retain the enum version from 3d188ab6a0fea01c43f48b5535ea096d7f287f9b, despite it not being as future-proof.
3. Use some kind of hack to work around the inability to have an optional dictionary, like a union type composed of either a unit type or an `AttributionReportingRequestOptions` dictionary.

Any other ideas?

**Comment by apasel422** [comment]
Based on what Trust Tokens is doing, it seems that for `fetch` we can [make the dictionary non-nullable on `RequestInit`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/fetch/request_init.idl;l=32;drc=71fadf4673bc87f9a9877eb457ae4adb118670e9) but [dynamically check for its presence](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/fetch/request.cc;l=578;drc=9d9593d7fc3fa8fa70dfd4dc7f5c98dc8b2dbc84).

For XHR we can use [a setter](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/xmlhttprequest/xml_http_request.idl;l=63;drc=71fadf4673bc87f9a9877eb457ae4adb118670e9).

**Comment by csharrison** [comment]
> Is XHR a strong use case? Can't folks that care about this move to `fetch()`? (I'd have assumed most already use `fetch()`, but I could be wrong)

We can ask for feedback, but previously when we tried to omit XHR support from Private State Tokens we heard very strong feedback that it would require major refactoring for partners and they were not interested in moving to `fetch` simultaneously with testing the new API. cc @dvorak42. Note that but PST and ARA integrate tightly with existing legacy flows.

**Comment by jfggit** [comment]
A few questions:

1. Just to check that I'm reading it correctly, would basically look like this? (And are both event and trigger properties actually required to be present?):

```
const attributionOptions = {eventSourceEligible: true, triggerEligible: false};
window.fetch('https://view.url', {attributionReporting: attributionOptions});
```
2. Does this imply you could register trigger event sources from fetch? Would there be any kind of requirement like "happens close to a user-driven interaction event" or anything? I could see that being helpful for registering things like third-party click trackers.

3. What would the estimated timeline look like, and what does the switchover look like? Is it just once the newer release goes out then setting the header is broken and the new option is supported? I assume it would be safe for fetch clients to just do both (i.e. set header and set options) in the interim?

**Comment by apasel422** [comment]
> 1. Just to check that I'm reading it correctly, would basically look like this? (And are both event and trigger properties actually required to be present?):
> 
> ```
> const attributionOptions = {eventSourceEligible: true, triggerEligible: false};
> window.fetch('https://view.url', {attributionReporting: attributionOptions});
> ```

Yes, as shown in `EVENT.md`. If you want to register sources via `fetch` then both options need to be specified. 

> 2. Does this imply you could register trigger event sources from fetch? Would there be any kind of requirement like "happens close to a user-driven interaction event" or anything? I could see that being helpful for registering things like third-party click trackers.

You can already register event sources from `fetch` today, by setting the `Attribution-Reporting-Eligible: event-source` request header in the options parameter. If you look at the commit diffs, you'll see the current way of doing this. This change just makes doing so supported in a more first-class way. There's no interaction requirements with this change or with the existing behavior.

> 3. What would the estimated timeline look like, and what does the switchover look like? Is it just once the newer release goes out then setting the header is broken and the new option is supported? I assume it would be safe for fetch clients to just do both (i.e. set header and set options) in the interim?

TBD, but I assume that there will be an atomic switchover, in which case it is safe to do both, though you could also feature-detect this with something like ` if ('attributionReporting' in XMLHttpRequest.prototype)`.

**Comment by apasel422** [comment]
> Re: 1 - oh, looking at EVENT.md now, don't the new examples each only set one of the two properties? Is that right?

As demonstrated in `EVENT.md`, both properties need to be *present* in the dictionary, but they can contain an arbitrary mixture of `true`/`false`.

**Comment by csharrison** [comment]
> This change seems fine to make to me, if a little late in the process. In any case, it would be good to have the API locked down and an estimate of the timing as soon as possible just to minimize additional complications.

Hey @jfggit , I agree w/ you we want to get the API locked down ASAP given that we want to ship this soon. Unfortunately the feedback from issue #756 came quite recently so we are only able to respond to this now.

In general, during Chromium's Origin Trial we want to make sure we can make changes like this in response to feedback where possible; that's the reason we do OTs in the first place to enable faster iteration on APIs. Once we graduate past OT, we will be much more careful about backwards incompatible changes (in fact, at least in Chromium they will require independent approvals via the intent process).

**Comment by jfggit** [comment]
+1, agreed with the comments in #756 about consistency. And actually, per that comment, do we expect the CORS requirement to remain after this change?

**Comment by apasel422** [comment]
> And actually, per that comment, do we expect the CORS requirement to remain after this change?

I'm not completely sure, but I suspect CORS will no longer be necessary.

**Comment by jfggit** [comment]
> I'm not completely sure, but I suspect CORS will no longer be necessary.

Great, it would good to get clarity on this whenever possible. That would be a potentially significant advantage over the status quo.

**Comment by apasel422** [comment]
> Great, it would good to get clarity on this whenever possible. That would be a potentially significant advantage over the status quo.

I will update this PR with information as I get it. We could have prioritized this long ago, but no one has mentioned the CORS concerns since this was first implemented in M105.

File: index.bs
Comment (yoavweiss) [inline index.bs:154]: This part should be a monkey patch on the Fetching processing model.
e.g. "In [http network or cache fetch](https://fetch.spec.whatwg.org/#http-network-or-cache-fetch) after step 8.14, call "Append attribution reporting", where "append attribution reporting" is an algorithm defined here which adds the header when conditions are met, and flips on a boolean that enabled trigger registration with the response.

Something similar should be done for XHR

Comment (apasel422) [reply index.bs:154]: Yes, we do need to integrate with `fetch` itself, but that's true even without this change.

Comment (yoavweiss) [reply index.bs:154]: Is that something you're interested in doing as a followup? IME, it's better to start out early (and small) and expand from there.

Comment (apasel422) [reply index.bs:154]: I plan to do it, but we also need to monkeypatch navigation (also even without this change), and I don't think doing so should block getting this landed as soon as possible, though I will take a look at least the fetch part now, while we're still waiting for feedback on this.

**Review by yoavweiss — APPROVED** [review]
LGTM

**Pull request merged**
